packages/nx/src/daemon/server/file-watching/changed-projects.ts

```diff
@@ -0,0 +1,71 @@
+import { performance } from 'perf_hooks';
+import { projectFileMapWithFiles } from '../project-graph-incremental-recomputation';
+
+export type ChangedFile = {
+  path: string;
+  type: 'CREATED' | 'UPDATED' | 'DELETED';
+};
+
+export function getProjectsAndGlobalChanges(
+  createdFiles: string[] | null,
+  updatedFiles: string[] | null,
+  deletedFiles: string[] | null
+) {
+  const projectAndGlobalChanges: {
+    projects: { [changedProject: string]: ChangedFile[] };
+    globalFiles: ChangedFile[];
+  } = {
+    projects: {},
+    globalFiles: [],
+  };
+
+  performance.mark('changed-projects:start');
+
+  const projectFileMap = projectFileMapWithFiles?.projectFileMap ?? {};
+
+  const allChangedFiles: ChangedFile[] = [
+    ...(createdFiles ?? []).map<ChangedFile>((c) => ({
+      path: c,
+      type: 'CREATED',
+    })),
+    ...(updatedFiles ?? []).map<ChangedFile>((c) => ({
+      path: c,
+      type: 'UPDATED',
+    })),
+    ...(deletedFiles ?? []).map<ChangedFile>((c) => ({
+      path: c,
+      type: 'DELETED',
+    })),
+  ];
+
+  for (const changedFile of allChangedFiles) {
```

Comment:
Do you know how slow this can be? Imagine if reformated 1000 files in a big workspace.

Currently it's O(Number of changes files * Number of files)

I think a better way might be to create a map from files to projects. You complexity will change to O(Number of files + number of changed files)